File: HW5/task1.cpp
#include <iostream>
#include <string>

int SIZE = 10e5;
int A = 7;
int PRIME = 1046527;
struct Node {
  int value;
  int empty;
  int rip;
};

class Set {
 private:
  Node **arr;

	bool exists(int value, int &i) {

		i = this->hash(value);
		while (!this->arr[i % SIZE]->empty) {
			if (this->arr[i % SIZE]->value == value && this->arr[i % SIZE]->rip == 0) {
				return (true);
			}
			i++;
		}
		return (false);
	}

  public:
	  Set() {
		  this->arr = (Node**)malloc(sizeof(Node*) * SIZE);
		  for (int i = 0; i < SIZE; i++) {
			this->arr[i] = (Node *)malloc(sizeof(Node));
			this->arr[i]->empty = 1;
			this->arr[i]->rip = 0;
		  }
	  }


  void insert (int value) {
	  int i;


	  if (exists(value, (int &) i)) {
		  return;
	  }
	  this->arr[i % SIZE]->value = value;
	  this->arr[i % SIZE]->empty = 0;
  }
  void del (int value) {
  	int i;

	  if (!exists(value, (int &) i)) {
		  return;
	  }
	  this->arr[i % SIZE]->rip = 1;
  }

  bool exists(int value) {
	  int i;

	  return exists(value, (int &) i);
  }

  int hash (int value) {
	return (A * value % PRIME) % SIZE;
  }

};


int main() {
	Set my_set = Set();
	int value;

	std::string str;
	while (std::cin >> str >> value) {
		if (str == "insert") {
			my_set.insert(value);
		}
		else if (str == "exists") {
			if (my_set.exists(value)) {
				std::cout << "true" << "\n";
			}
			else {
				std::cout << "false" << "\n";
			}
		}
		else if (str == "delete") {
			my_set.del(value);
		}
		if (std::cin.peek() == EOF) {
			break;
		}
	}
	return (0);
}

File: HW5/task1.py
import sys

PRIME = 3882959


class Set:

    def __init__(self, a, size):
        self.arr = ["e" for i in range(size)]
        self.rip = 0
        self.fill = 0
        self.a = a
        self.size = size

    def insert(self, value):

        self.rehash()
        exist, i = self.exist(value)
        if exist:
            return True
        self.arr[i] = value
        self.fill += 1

    def rehash(self):
        if self.rip + self.fill > 0.5 * self.size:
            if self.fill > 0.25 * self.size:
                self.a += self.fill % self.size
                self.size *= 2
                copy_arr = self.arr
                self.arr = ["e" for i in range(self.size)]
                self.fill = 0
                self.rip = 0
                for value in copy_arr:
                    if value != "r" and value != "e":
                        self.insert(value)
            else:
                self.a += self.fill % self.size
                self.size = self.size // 2
                copy_arr = self.arr
                self.fill = 0
                self.rip = 0
                self.arr = ["e" for i in range(self.size)]
                for value in copy_arr:
                    if value != "r" and value != "e":
                        self.insert(value)

    def delete(self, value):
        exist, i = self.exist(value)
        if not exist:
            return
        self.arr[i % self.size] = "r"
        self.rip += 1
        self.fill -= 1

    def hash(self, value):
        return (self.a * value % PRIME) % self.size

    def exist(self, value):
        i = self.hash(value)

        if self.arr[i] == value:
            return True, i
        while self.arr[i % self.size] != "e":
            if self.arr[i % self.size] == value:
                return True, i
            i = i + 1
        return False, i % self.size


if __name__ == "__main__":
    my_set = Set(7,1000)
    ans = []
    for line in sys.stdin.read().splitlines():
        if line.split()[0] == "insert":
            my_set.insert(int(line.split()[1]))
        elif line.split()[0] == "exists":
            if my_set.exist(int(line.split()[1]))[0]:
                ans.append("true")
            else:
                ans.append("false")
        elif line.split()[0] == "delete":
            my_set.delete(int(line.split()[1]))
    sys.stdout.buffer.write("\n".join(ans).encode())
